File: windows/Tokendial.Core/Model/UsageModel.cs
using Tokendial.Core.I18n;

namespace Tokendial.Core.Model;

/// <summary>Whether a number came from the vendor or was worked out locally. Derived readings are shown with a tilde and no dial arc.</summary>
public enum Fidelity
{
    Official,
    Derived
}

/// <summary>Colour band of a reading. Critical starts at 0.80 so the colour change and the 80 % alert coincide.</summary>
public enum Band
{
    Ample,
    Watch,
    Critical
}

public static class Bands
{
    public static Band Of(double usedFraction) => usedFraction switch
    {
        < 0.50 => Band.Ample,
        < 0.80 => Band.Watch,
        _ => Band.Critical
    };
}

/// <summary>How much to trust the reading on screen right now.</summary>
public abstract record ReadingStatus
{
    public sealed record Live : ReadingStatus;
    public sealed record Stale(DateTimeOffset Since) : ReadingStatus;
    public sealed record NeedsSignIn : ReadingStatus;
    public sealed record Unsupported(string Why) : ReadingStatus;
    public sealed record Failed(string Why) : ReadingStatus;

    public static readonly ReadingStatus LiveNow = new Live();
    public static readonly ReadingStatus SignIn = new NeedsSignIn();

    public bool IsStale => this is Stale;
    public DateTimeOffset? StaleSince => (this as Stale)?.Since;
}

/// <summary>One metered window: a fraction used, or a bare count when the vendor publishes no ceiling.</summary>
public sealed record UsageWindow(string Id, string Label, double? UsedFraction = null, int? Count = null, DateTimeOffset? ResetsAt = null)
{
    /// <summary>"63% used · 37% left", or "~7 requests today" for a count.</summary>
    public string Summary(Fidelity fidelity)
    {
        if (UsedFraction is double f)
        {
            var used = (int)Math.Round(f * 100, MidpointRounding.AwayFromZero);
            var tilde = fidelity == Fidelity.Derived ? "~" : "";
            return tilde + Strings.T("copy.usedLeft", ("used", used), ("left", Math.Max(0, 100 - used)));
        }
        if (Count is int n) return Strings.Plural("copy.requests", n);
        return Strings.T("copy.noReading");
    }
}

/// <summary>Set when a limit has been reached even where the headline still shows room.</summary>
public sealed record Blocked(string Reason, DateTimeOffset? Until);

/// <summary>Everything the panel shows for one provider.</summary>
public sealed record ProviderReading(
    string ProviderId,
    string DisplayName,
    Fidelity Fidelity,
    ReadingStatus Status,
    IReadOnlyList<UsageWindow> Windows,
    string? HeadlineId = null,
    Blocked? Block = null)
{
    /// <summary>The window the dial means. Declared by the provider; when it is missing the dial shows no reading rather than another window.</summary>
    public UsageWindow? Headline => HeadlineId is null ? Windows.FirstOrDefault() : Windows.FirstOrDefault(w => w.Id == HeadlineId);

    public double? HeadlineFraction => Headline?.UsedFraction;

    public bool HasReading => Windows.Count > 0;

    /// <summary>"63%", a bare count, or a dash.</summary>
    public string HeadlineText =>
        HeadlineFraction is double f ? $"{(int)Math.Round(f * 100, MidpointRounding.AwayFromZero)}%"
        : Headline?.Count is int n ? n.ToString()
        : "—";

    public Band? Band => Block is not null ? Model.Band.Critical : HeadlineFraction is double f ? Bands.Of(f) : null;
}

File: macos/TokendialCore/Sources/TokendialCore/Model/UsageModel.swift
import Foundation

/// Whether a number came from the vendor or was worked out locally. Derived readings are shown with a tilde and no dial arc.
public enum Fidelity: String, Codable, Sendable {
    case official
    case derived
}

/// Colour band of a reading. Critical starts at 0.80 so the colour change and the 80 % alert coincide.
public enum Band: Sendable {
    case ample
    case watch
    case critical

    public static func of(_ usedFraction: Double) -> Band {
        if usedFraction < 0.50 { return .ample }
        if usedFraction < 0.80 { return .watch }
        return .critical
    }
}

/// How much to trust the reading on screen right now.
public enum ReadingStatus: Equatable, Codable, Sendable {
    case live
    case stale(since: Date)
    case needsSignIn
    case unsupported(why: String)
    case failed(why: String)

    public var isStale: Bool { if case .stale = self { return true } else { return false } }
    public var staleSince: Date? { if case .stale(let since) = self { return since } else { return nil } }
}

/// One metered window: a fraction used, or a bare count when the vendor publishes no ceiling.
public struct UsageWindow: Equatable, Codable, Sendable {
    public var id: String
    public var label: String
    public var usedFraction: Double?
    public var count: Int?
    public var resetsAt: Date?

    public init(id: String, label: String, usedFraction: Double? = nil, count: Int? = nil, resetsAt: Date? = nil) {
        self.id = id
        self.label = label
        self.usedFraction = usedFraction
        self.count = count
        self.resetsAt = resetsAt
    }

    /// "63% used · 37% left", or "~7 requests today" for a count.
    public func summary(_ fidelity: Fidelity) -> String {
        if let f = usedFraction {
            let used = Int((f * 100).rounded(.toNearestOrAwayFromZero))
            let tilde = fidelity == .derived ? "~" : ""
            return tilde + Strings.t("copy.usedLeft", ["used": used, "left": max(0, 100 - used)])
        }
        if let n = count { return Strings.plural("copy.requests", n) }
        return Strings.t("copy.noReading")
    }
}

/// Set when a limit has been reached even where the headline still shows room.
public struct Blocked: Equatable, Codable, Sendable {
    public var reason: String
    public var until: Date?

    public init(reason: String, until: Date?) {
        self.reason = reason
        self.until = until
    }
}

/// Everything the panel shows for one provider.
public struct ProviderReading: Equatable, Codable, Sendable {
    public var providerId: String
    public var displayName: String
    public var fidelity: Fidelity
    public var status: ReadingStatus
    public var windows: [UsageWindow]
    public var headlineId: String?
    public var block: Blocked?

    public init(providerId: String, displayName: String, fidelity: Fidelity, status: ReadingStatus, windows: [UsageWindow], headlineId: String? = nil, block: Blocked? = nil) {
        self.providerId = providerId
        self.displayName = displayName
        self.fidelity = fidelity
        self.status = status
        self.windows = windows
        self.headlineId = headlineId
        self.block = block
    }

    /// The window the dial means. Declared by the provider; when it is missing the dial shows no reading rather than another window.
    public var headline: UsageWindow? {
        guard let id = headlineId else { return windows.first }
        return windows.first { $0.id == id }
    }

    public var headlineFraction: Double? { headline?.usedFraction }
    public var hasReading: Bool { !windows.isEmpty }

    /// "63%", a bare count, or a dash.
    public var headlineText: String {
        if let f = headlineFraction { return "\(Int((f * 100).rounded(.toNearestOrAwayFromZero)))%" }
        if let n = headline?.count { return String(n) }
        return "—"
    }

    public var band: Band? {
        if block != nil { return .critical }
        if let f = headlineFraction { return Band.of(f) }
        return nil
    }
}
